### PR TITLE
feat: add optional prompt storage support

### PR DESCRIPTION
**Pull request by bdura**
## Description

This PR adds prompt storage support. Prompt and responses are optionally saved to `doc._.llm_io`

### Types of change

This is a feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran all tests , and all new and existing tests passed. This includes tests marked as `external` and those 
  requiring a GPU. I did this for the following directories:
  - [x] `tests`
  - [x] `usage_example/tests`
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy_llm/backends/rest/backend/noop.py
Comment (rmitsch) [inline spacy_llm/backends/rest/backend/noop.py:6]: Let's make this a static class variable, if there's no particular reason to have it global.

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [inline spacy_llm/pipeline/llm.py:254]: Why the list conversion?

File: spacy_llm/tasks/noop.py
Comment (rmitsch) [inline spacy_llm/tasks/noop.py:7]: Variables not needed outside of the module (ignoring tests) should be prefixed with `_`.

**Review by rmitsch — COMMENTED** [review]
Looks good. Docs should be updated too.

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py:254]: Otherwise the iterator is consumed by `self._backend(prompts)`. Unless using `iter` makes it unnecessary?

File: spacy_llm/tasks/noop.py
Comment (bdura) [reply spacy_llm/tasks/noop.py]: Done.

File: spacy_llm/pipeline/llm.py
Comment (adrianeboyd) [reply spacy_llm/pipeline/llm.py:148]: I think what you're looking for might be `itertools.tee`?

Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:148]: Ah, you're using the prompts and responses for the logging again. Give it a shot with `itertools.tee()`?

Comment (bdura) [reply spacy_llm/pipeline/llm.py:148]: Since they are consumed sequentially [I'm not sure it will help](https://docs.python.org/3/library/itertools.html#itertools.tee:~:text=This%20itertool%20may%20require%20significant%20auxiliary%20storage%20(depending%20on%20how%20much%20temporary%20data%20needs%20to%20be%20stored).%20In%20general%2C%20if%20one%20iterator%20uses%20most%20or%20all%20of%20the%20data%20before%20another%20iterator%20starts%2C%20it%20is%20faster%20to%20use%20list()%20instead%20of%20tee().)...

Comment (adrianeboyd) [reply spacy_llm/pipeline/llm.py:148]: It's not that this doesn't work, but I think it's more confusing at first glance than using `itertools.tee`.

I do worry in general that this whole collection of multiple interleaved steps with iterators is very brittle and likely to get out-of-sync.

Comment (adrianeboyd) [reply spacy_llm/pipeline/llm.py:148]: It might also be clearer if the second iterators are only created if required?

Comment (bdura) [reply spacy_llm/pipeline/llm.py:148]: I agree. We could also defer the responsibility to the task... I did it this way because it does not require any change on the task, but come to think of it it would make sense that they handle it.

Consider the case where the `Prompt` is a complex, non-serialisable object... It should be handled differently.

Comment (bdura) [reply spacy_llm/pipeline/llm.py:148]: Ah, and I also struggled to include it at the task level _and_ get the component name...

Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:148]: > Consider the case where the Prompt is a complex, non-serialisable object... It should be handled differently.

We could only store the stringified version of prompts and responses?

Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:148]: > Ah, and I also struggled to include it at the task level and get the component name...

Can you elaborate on this?

Comment (bdura) [reply spacy_llm/pipeline/llm.py]: > We could only store the stringified version of prompts and responses?

Good idea. Then it's the task's/backends responsibility to generate prompts/responses that display a proper `__str__` method.

Comment (bdura) [reply spacy_llm/pipeline/llm.py]: I looked at a potential solution to implement prompt/response storage at the task level, but it looks very convoluted. We'd likely need to apply substantial changes to the architecture.

> Ah, and I also struggled to include it at the task level and get the component name...

Storage is done on a per-component basis so we need the component name. We could use a list or more exotic workarounds, but I think that's more confusing than anything else. Do you have another idea?

File: spacy_llm/pipeline/llm.py
Comment (adrianeboyd) [inline spacy_llm/pipeline/llm.py:150]: Not with this naming, but what I meant was more like:

```suggestion
                prompts = tee(self._task.generate_prompts(noncached_doc_batch))
                if self._save_io:
                    prompts, prompts_for_saving = tee(prompts)
                responses = self._backend(prompts))
                if self._save_io:
                    responses, responses_for_saving = tee(responses)
```

File: spacy_llm/pipeline/llm.py
Comment (adrianeboyd) [inline spacy_llm/pipeline/llm.py:151]: `prompts` is already consumed at this point?

File: spacy_llm/pipeline/llm.py
Comment (adrianeboyd) [inline spacy_llm/pipeline/llm.py:162]: And then you don't need this.

Comment (bdura) [reply spacy_llm/pipeline/llm.py]: Right, I saw that... I pushed the necessary commits.

File: spacy_llm/pipeline/llm.py
Comment (adrianeboyd) [inline spacy_llm/pipeline/llm.py:149]: This is very confusing naming.

Comment (bdura) [reply spacy_llm/pipeline/llm.py]: Agreed. `saved_prompts` ?

File: README.md
Comment (adrianeboyd) [inline README.md:216]: I think the README could use a table of `llm` config settings?

Comment (bdura) [reply README.md:226]: I agree, it's overdue. I added a table and kept the two main settings apart like before, it seems important. Let me know what you think.

File: README.md
Comment (adrianeboyd) [inline README.md:210]: Ah, `name` shouldn't be one of the config settings. Let me check this in the refactoring...

Comment (adrianeboyd) [reply README.md:210]: It's fine, just `nlp` and `name` aren't in the config as config settings.

```suggestion
```

Comment (bdura) [reply README.md]: Thanks !

File: spacy_llm/pipeline/llm.py
Comment (kabirkhan) [inline spacy_llm/pipeline/llm.py:137]: This needs to be added to `__call__` as well. Since our backend + response parser always work on batches I think putting this in a util method would work best. 

We're deferring this logging PR (#80) for 0.3.0 but the `_process_docs` method could be added to this PR I think. Then the extension installation (+ setting of the prompt/response) only needs to happen in one spot.

[_process_docs method](https://github.com/explosion/spacy-llm/blob/3e7c481a0300f3fd227df98b19785373305c3c81/spacy_llm/pipeline/llm.py#L242)

Comment (bdura) [reply spacy_llm/pipeline/llm.py:137]: Ah absolutely, thanks! I'll wait for your PR to be merged then.

**Comment by bdura** [comment]
Waiting for #80 to be merged.

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [reply spacy_llm/pipeline/llm.py]: Ah, ok. I think it's fine to do this at the component level. I don't think customized logging (via tasks) is a priority right now.

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:137]: I agree with Kabir, but I don't think we need to wait for https://github.com/explosion/spacy-llm/pull/80 to be merged. I'd recommend copying over `_process_docs()` and adding setting the extension.

Comment (kabirkhan) [reply spacy_llm/pipeline/llm.py:137]: Oh yeah, that's what I meant, I think this PR should be merged first but we should copy over the core of `_process_docs` so we only need to install the extensions in 1 method.

File: spacy_llm/pipeline/llm.py
Comment (kabirkhan) [inline spacy_llm/pipeline/llm.py:149]: I don't think `itertools.tee` helps here since we consume the iterator in the backend but I've never properly profiled this so I'm not sure

File: spacy_llm/pipeline/llm.py
Comment (kabirkhan) [inline spacy_llm/pipeline/llm.py:41]: I personally think we should default this to True. Is memory consumption/storage space the only concern here?

**Comment by kabirkhan** [comment]
See https://github.com/explosion/spacy-llm/pull/147 for refactored `_process_docs` method so we only need to install extension attrs there, not separately in `__call__` and `pipe`

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:41]: I think so. I'm on the fence about whether to make this opt-out.

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [reply spacy_llm/pipeline/llm.py:149]: Performance isn't that critical here anyway, I believe.

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py:137]: Alright!

**Comment by adrianeboyd** [comment]
One concern is what happens if this is incorporated into spacy at some point: spacy itself has a policy of never setting any custom extensions. There are a few tokenizers that add data to `doc.user_data` (the same place extension data is stored), but not in a way that interacts with custom extensions data. The idea is that these might conflict with user-provided extensions so it's better if spacy never conflicts with any potential user code here.

**Comment by rmitsch** [comment]
https://github.com/explosion/spacy-llm/pull/147 has been merged (but you already saw that).

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py:129]: Done, thanks.

**Comment by bdura** [comment]
> One concern is what happens...

I changed it to use `doc.user_data` directly. Let me know.

@rmitsch there's still the opt-in/out decision to make. Otherwise I think we're good to go.

**Comment by rmitsch** [comment]
I'm inclined to leave it as `False`. We can also change that in a later PR, if we feel that's better.

File: spacy_llm/pipeline/llm.py
Comment (rmitsch) [inline spacy_llm/pipeline/llm.py:165]: Should be removed, no?

**Review by rmitsch — COMMENTED** [review]
LGTM for me except for the commented code. I'd want @adrianeboyd to sign off on storing this in `user_data` before we merge though.

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py:165]: Gosh. Thanks

Comment (bdura) [reply spacy_llm/pipeline/llm.py]: Done.

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py:41]: It will be opt-in for now (see [comment](https://github.com/explosion/spacy-llm/pull/127#issuecomment-1569706839))

File: spacy_llm/pipeline/llm.py
Comment (bdura) [reply spacy_llm/pipeline/llm.py]: Exactly. See [comment](https://github.com/explosion/spacy-llm/pull/127#discussion_r1206379607).

**Comment by adrianeboyd** [comment]
So the only places does this are pretty obscure, just the Japanese and Korean tokenizers, but I think this is okay.

It is kind internals, so I think it would make sense to provide utility functions (+ usage examples) that can be used to access this data with custom extension getters. So you could do something like:

```python
doc._.llm_io(pipe_name)["prompt"]
```

or

```python
doc._.llm_io(pipe_name).prompt
```

**Comment by bdura** [comment]
@adrianeboyd in that case maybe using `doc._.llm_io` directly makes more sense for now (since it's conceptually simpler)? If we're going to use a custom extension in both cases...

**Comment by rmitsch** [comment]
@adrianeboyd Sorry, it's not quite clear to me what you are recommending. Should we use `user_data` or `_.llm_io`?

**Comment by adrianeboyd** [comment]
Let's stick with `user_data` for this PR. The idea with custom extensions would be that users would have to run some util method to add the custom extensions, just for easier access to this data, and it would be easy to add something like that later.

**Comment by bdura** [comment]
I pushed a change to remove any mention of `Doc._.llm_io`. I think we're good to go, although if you have a better suggestion regarding the naming conventions I'd love your input.

**Comment by bdura** [comment]
By the way @rmitsch: watch out, this is set up to merge into main. There's nothing breaking though, so we should be fine – just thought I'd mention it nonetheless.

**Comment by adrianeboyd** [comment]
Since this is adding a config setting to `llm`, I think it should go in the next minor version.

**Comment by adrianeboyd** [comment]
Ah, too many differences `develop` to switch just in the UI here, sorry.

**Comment by rmitsch** [comment]
@bdura Can you target this PR at `develop` and resolve the conflicts?

**Review by rmitsch — APPROVED** [review]
Great work! Merging after the tests pass.

**Comment by bdura** [comment]
Give me a minute :)

**Pull request merged**
